### ISSUE TITLE
[a11y]: Out-of-order keyboard navigation when MenuItems added dynamically

### ISSUE THREAD

**Issue by wkeese** [carbon-design-system/carbon#20004]
### Package

@carbon/react

### Browser

Chrome

### Operating System

MacOS

### Package version

1.83

### React version

18

### Automated testing tool and ruleset

manual testing

### Assistive technology

none

### Description

# Background

React lets the application dynamically add/remove components.  So with code like this:

```jsx
<Menu>
        {items.map((item) => (
          <MenuItem key={item} label={item} />
        ))}
</Menu>
```

If you add a new item somewhere in the `items[]` array, React will create a new `MenuItem` component (i.e. a new `<li>` node) without disturbing the pre-existing ones.

# Problem

However, that dynamically added `MenuItem` breaks keyboard navigation.  Specifically, even if the dynamically added `MenuItem` is the first item in the `Menu`, it becomes the last navigable item with the down arrow key.

# Cause

In Menu.tsx, keyboard navigation is based on the order of the items in the `focusableItems[]` array, and `focusableItems[]` is based on `childContext.state.items[]`.

However, the order of the items in `childContext.state.items` is based on the order the `MenuItems` were *created*, not their logical order in the DOM.   Because each MenuItem registers itself when it's created, adding itself to the end of the `childContext.state.items[]` array.

### Reproduction/example

https://stackblitz.com/edit/github-ykbrpwcg-8pvggqsy?file=src%2FExample%2FExample.jsx

### Steps to reproduce

Just try pressing the "add menuitem" button and then navigation through the menu using up/down arrows.

Actually my test case works a little differently from my actual app.   In my actual app, the newly added MenuItem is last in the up/down arrow navigation, even though it's the second MenuItem in the list.  Not sure why the discrepancy.

### Suggested Severity

None

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [x] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

**Comment by Mikadv** [comment]
Saw: Using the down key on my keyboard doesn't work if I am on `0`. But clicking the up key does work.
